### PR TITLE
Standard MasterChooser revamping

### PR DESCRIPTION
**Pull request by stratomda**
Addresses the following tasks in #257 

- [x] Catching exception when master URI is a valid IP but no roscore is running. Currently in indigo the application crashes 
- [X] Catch malformed URI 
- [X] Add a default port if it is missing

### CODE REVIEW

File: android_10/src/org/ros/android/MasterChooser.java [hunk 1]
```diff
@@ -111,16 +131,22 @@ public boolean hasStableIds() {
...
     connectButton = (Button) findViewById(R.id.master_chooser_ok);
     uriText.addTextChangedListener(new TextWatcher() {
       @Override
       public void onTextChanged(CharSequence s, int start, int before, int count) {
-        if (s.length() > 0) {
-          connectButton.setEnabled(true);
-        } else {
-          connectButton.setEnabled(false);
-        }
+          final String uri = s.toString();
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: very minor nit: I think there is an indentation issue here, right? Shouldn't it be two spaces? (to follow the rest of the repository's convention)

File: android_10/src/org/ros/android/MasterChooser.java [hunk 2]
```diff
@@ -181,10 +207,25 @@ public void onActivityResult(int requestCode, int resultCode, Intent intent) {
   }
 
   public void okButtonClicked(View unused) {
+    String tmpURI = uriText.getText().toString();
+
+    //Check to see if the URI has a port.
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: truly minor nit: add a space between // and the text

Comment (adamantivm) [reply android_10/src/org/ros/android/MasterChooser.java]: same elsewhere

File: android_10/src/org/ros/android/MasterChooser.java [hunk 3]
```diff
@@ -297,4 +349,88 @@ protected boolean isQRCodeReaderInstalled(Intent intent) {
...
+            + "(?:"
+            + "(?:" + HTTP_PROTOCOL + ")"
+            + "(?:" + RELAXED_DOMAIN_NAME + ")"
+            + "(?:" + PORT_NUMBER + ")?"
+            + ")"
+            + WORD_BOUNDARY
+            + ")");
+
+    public static final Pattern PORT = Pattern.compile(PORT_NUMBER);
+
```
Comment (adamantivm) [inline android_10/src/org/ros/android/MasterChooser.java]: nit, extra line

**Review by adamantivm — APPROVED** [review]
@stratomda thank you very much for this, great work! I left only a few very minor nits about formatting. If you could please address them I'll merge right away.

File: android_10/src/org/ros/android/MasterChooser.java [hunk 4]
```diff
@@ -111,16 +131,22 @@ public boolean hasStableIds() {
...
     uriText.addTextChangedListener(new TextWatcher() {
       @Override
       public void onTextChanged(CharSequence s, int start, int before, int count) {
-        if (s.length() > 0) {
-          connectButton.setEnabled(true);
-        } else {
-          connectButton.setEnabled(false);
-        }
+          final String uri = s.toString();
+          if(!uriPattern.matcher(uri).matches())
```
Comment (jubeira) [inline android_10/src/org/ros/android/MasterChooser.java]: micro nit regarding coding style:
```java
if (!uriPattern.matcher(uri).matches()) {
```

File: android_10/src/org/ros/android/MasterChooser.java [hunk 5]
```diff
@@ -181,10 +207,25 @@ public void onActivityResult(int requestCode, int resultCode, Intent intent) {
   }
 
   public void okButtonClicked(View unused) {
+    String tmpURI = uriText.getText().toString();
+
+    //Check to see if the URI has a port.
+    final Pattern portPattern = RosURIPattern.PORT;
+    if(!portPattern.matcher(tmpURI).find())
```
Comment (jubeira) [inline android_10/src/org/ros/android/MasterChooser.java]: same as above: 
```java
if (!portPattern.matcher(tmpURI).find()) {
```

File: android_10/src/org/ros/android/MasterChooser.java [hunk 6]
```diff
@@ -204,6 +245,17 @@ protected Boolean doInBackground(Void... params) {
...
+        {
+          String exceptionMessage = e.getMessage();
+          if(exceptionMessage.contains(CONNECTION_EXCEPTION_TEXT))
+            toast("Unable to communicate with master!");
+          else if(exceptionMessage.contains(UNKNOW_HOST_TEXT))
+            toast("Unable to resolve URI hostname!");
+          else
+            toast("Communication error!");
+          return false;
+        }
```
Comment (jubeira) [inline android_10/src/org/ros/android/MasterChooser.java:255]: and one more:
```java
       catch (Exception e) {
         String exceptionMessage = e.getMessage();
         if (exceptionMessage.contains(CONNECTION_EXCEPTION_TEXT))
           toast("Unable to communicate with master!");
         else if (exceptionMessage.contains(UNKNOW_HOST_TEXT))
           toast("Unable to resolve URI hostname!");
         else
           toast("Communication error!");
         return false;
       }
```

**Review by jubeira — APPROVED** [review]
Looks good, great contribution! Thanks @stratomda !
I only added a few really minor comments regarding code style.

**Pull request merged**